### PR TITLE
Fix and refactor layer merging and evaluation

### PR DESCRIPTION
**Pull request by mfehr**
I wanted to test the layer merging and add some more detailed layer-to-layer evaluation function, which we need for modelify. What I ended up doing in the end:
 * Complete the SDF to PLY conversion function in sdf_ply.h
     * It now supports Tsdf and Esdf and the IsoSurface mode outputs the surface mesh
 * As a consequence the MeshIntegrator needed to support EsdfLayers as well (it was templated, but accessed TsdfVoxel fields inside, so it wouldn't have compiled if someone used it with Esdf (?))
 * Remove the MergeIntegrator class and turned it into free functions, because it had no state anyways.
 * Fix bug in Interpolator (left: before, right: after, bottom: ground truth)
image
image

 * Add a end to end unit test for layer merging
 * Move evaluation tools from simulation server to their own headers
 * Add more detailed evaluation options (e.g. include # overlapping voxels)

### CODE REVIEW

File: voxblox/include/voxblox/utils/evaluation_utils.h [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+#include "voxblox/core/layer.h"
+#include "voxblox/core/voxel.h"
+
+namespace voxblox {
+namespace utils {
+
+enum class VoxelEvaluationResult { kNoOverlap, kIgnored, kEvaluated };
+
+enum class VoxelEvaluationMode {
```
Comment (helenol) [inline voxblox/include/voxblox/utils/evaluation_utils.h]: Nice :) I like this solution.

File: voxblox/include/voxblox/utils/evaluation_utils.h [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  BlockIndexList block_list;
+  layer_test.getAllAllocatedBlocks(&block_list);
+  size_t vps = layer_test.voxels_per_side();
+  size_t num_voxels_per_block = vps * vps * vps;
+
+  VoxelEvaluationDetails evaluation_details;
+
+  double total_squared_error = 0.0;
+
+  for (const BlockIndex block_index : block_list) {
```
Comment (helenol) [inline voxblox/include/voxblox/utils/evaluation_utils.h]: Also ref? Any reason why not?

File: voxblox/include/voxblox/utils/evaluation_utils.h [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        default:
+          LOG(FATAL) << "Unkown voxel evaluation result: "
+                     << static_cast<int>(result);
+      }
+    }
+  }
+
+  // Iterate over all blocks in the grond truth layer and look them up in the
+  // test truth layer. This is only done to get the exact number of
+  // non-verlapping voxels.
```
Comment (helenol) [inline voxblox/include/voxblox/utils/evaluation_utils.h]: non-overlapping

File: voxblox/include/voxblox/integrator/merge_integration.h [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  }
+
+  // get inverse transform
+  Transformation T_out_in = T_in_out.inverse();
+
+  Interpolator<VoxelType> interpolator(&layer_in);
+
+  // we now go through all the blocks in the output layer and interpolate the
+  // input layer at the center of each output voxel position
+  for (BlockIndex block_idx : block_idx_set) {
```
[the remark this answers is not in the document]
Comment (mfehr) [reply voxblox/include/voxblox/integrator/merge_integration.h]: nope, just code that was moved

**Pull request merged**